torch/testing/_internal/common_methods_invocations.py

```diff
@@ -328,21 +341,67 @@ def get_operator_variant(self):
         """
         return self.operator_variant
 
-    def sample_inputs(self, device, dtype, requires_grad=False, **kwargs):
-        """Returns an iterable of SampleInputs.
+    def generate_sample_inputs(self, device, dtype, requires_grad=False, **kwargs):
+        """Returns a list of SampleInputs.
 
         These samples should be sufficient to test the function works correctly
         with autograd, TorchScript, etc.
         """
-
         # TODO: Remove the try/except once all operators have sample_inputs_func with
         #       **kwargs in their signature.
         try:
             samples = self.sample_inputs_func(self, device, dtype, requires_grad, **kwargs)
         except TypeError:
             samples = self.sample_inputs_func(self, device, dtype, requires_grad)
+        if len(samples) == 0:
+            raise ValueError('sample_inputs_func must return at least 1 sample')
         return samples
 
+    def sample_inputs(self, device, dtype, requires_grad=False, **kwargs):
+        """Generates a sequence of SampleInputs.
+
+        The OpInfo keeps track the the sample which is currently yielded: wrap
+        your code inside a with op.decorate_errors() to get better exception
+        messages which include the current sample.
+
+        If subclasses want to customize the generation of inputs, they should
+        override generate_sample_inputs(). On the contrary, end users should
+        always call sample_inputs().
+        """
+        if self._current_sample is not None:
+            raise ValueError('Cannot call op.sample_inputs() while another '
+                             'iteration is in progress')
+        samples = self.generate_sample_inputs(device, dtype, requires_grad, **kwargs)
+        for i, sample in enumerate(samples):
+            assert isinstance(sample, SampleInput)
+            # make sure that every SampleInput has a sensible name: if the
+            # name was not given, use the index instead
+            if not sample.name:
+                sample.name = str(i)
+            self._current_sample = sample
+            yield sample
+        self._current_sample = None
```

Comment:
I imagine this reference is why the tests are failing with CUDA memory leaks (because it's keeping the samples alive after the test's conclusion)

Reply:
I think you are right, but it's an easy fix. I pushed d05a010730 which fixes the tests on my machine, let's see what happens on buildbots